### PR TITLE
Respond to RangeHelper refactor

### PR DESCRIPTION
**Pull request by jbagga**
Reaction PR for https://github.com/aspnet/StaticFiles/pull/200
Issue: https://github.com/aspnet/StaticFiles/issues/196

Behavior now closely matches StaticFiles for invalid ranges (as a result, some changes were made to tests)

cc @Tratcher

### CODE REVIEW

File: src/Microsoft.AspNetCore.Mvc.Core/Internal/FileResultExecutorBase.cs [hunk 1]
```diff
@@ -246,16 +245,22 @@ private static PreconditionState GetMaxPreconditionState(params PreconditionStat
...
-            EntityTagHeaderValue etag = null)
+            long? fileLength)
         {
             var response = context.HttpContext.Response;
             var httpResponseHeaders = response.GetTypedHeaders();
 
             // Checked for presence of Range header explicitly before calling this method.
             // Range may be null for parsing errors, multiple ranges and when the file length is missing.
-            var range = fileLength.HasValue ? ParseRange(context, httpRequestHeaders, fileLength.Value, lastModified, etag) : null;
+            var (isRangeRequest, range) = ParseRange(
```
Comment (Tratcher) [inline src/Microsoft.AspNetCore.Mvc.Core/Internal/FileResultExecutorBase.cs]: Why attempt range processing for files without lengths?

File: src/Microsoft.AspNetCore.Mvc.Core/Internal/FileResultExecutorBase.cs [hunk 2]
```diff
@@ -290,30 +295,15 @@ private static long SetContentLength(ActionContext context, RangeItemHeaderValue
...
+        private static (bool, RangeItemHeaderValue) ParseRange(
             ActionContext context,
             RequestHeaders httpRequestHeaders,
-            long fileLength,
-            DateTimeOffset? lastModified = null,
-            EntityTagHeaderValue etag = null)
+            long? fileLength)
         {
             var httpContext = context.HttpContext;
             var response = httpContext.Response;
```
Comment (Tratcher) [inline src/Microsoft.AspNetCore.Mvc.Core/Internal/FileResultExecutorBase.cs]: unused. Not much point in keeping this whole method anymore.

File: src/Microsoft.AspNetCore.Mvc.Core/Internal/FileResultExecutorBase.cs [hunk 3]
```diff
@@ -79,11 +79,10 @@ internal enum PreconditionState
             if (HttpMethods.IsHead(request.Method) || HttpMethods.IsGet(request.Method))
             {
                 var preconditionState = GetPreconditionState(context, httpRequestHeaders, lastModified, etag);
-                if (request.Headers.ContainsKey(HeaderNames.Range) &&
-                    (preconditionState == PreconditionState.Unspecified ||
+                if ((preconditionState == PreconditionState.Unspecified ||
```
Comment (Tratcher) [inline src/Microsoft.AspNetCore.Mvc.Core/Internal/FileResultExecutorBase.cs]: Are you going to get rid of PreconditionState.IgnoreRangeRequest in MVC too?

Comment (jbagga) [inline src/Microsoft.AspNetCore.Mvc.Core/Internal/FileResultExecutorBase.cs]: I'll need to introduce _isRangeRequest or something similar to check if the range should be served anyway. Any drawbacks for using this?

Comment (jbagga) [inline src/Microsoft.AspNetCore.Mvc.Core/Internal/FileResultExecutorBase.cs]: I have refactored it to check if the IfRange header satisfies the conditions before calling SetRangeHeader

**Comment by jbagga** [comment]
@Tratcher Updated

File: src/Microsoft.AspNetCore.Mvc.Core/Internal/FileResultExecutorBase.cs (hunk 3 above)
Comment (Tratcher) [inline src/Microsoft.AspNetCore.Mvc.Core/Internal/FileResultExecutorBase.cs]: Avoid divergent designs, they're hard to maintain.

File: src/Microsoft.AspNetCore.Mvc.Core/Internal/FileResultExecutorBase.cs [hunk 4]
```diff
@@ -80,24 +70,35 @@ internal enum PreconditionState
...
-                    return SetRangeHeaders(context, httpRequestHeaders, fileLength, lastModified, etag);
-                }
-                if (preconditionState == PreconditionState.NotModified)
-                {
-                    serveBody = false;
-                    response.StatusCode = StatusCodes.Status304NotModified;
-                }
-                else if (preconditionState == PreconditionState.PreconditionFailed)
+                serveBody = false;
+                response.StatusCode = StatusCodes.Status304NotModified;
```
Comment (Tratcher) [inline src/Microsoft.AspNetCore.Mvc.Core/Internal/FileResultExecutorBase.cs]: return?

File: src/Microsoft.AspNetCore.Mvc.Core/Internal/FileResultExecutorBase.cs [hunk 5]
```diff
@@ -80,24 +70,35 @@ internal enum PreconditionState
...
+            }
+            else if (preconditionState == PreconditionState.PreconditionFailed)
+            {
+                serveBody = false;
+                response.StatusCode = StatusCodes.Status412PreconditionFailed;
```
Comment (Tratcher) [inline src/Microsoft.AspNetCore.Mvc.Core/Internal/FileResultExecutorBase.cs]: return?

File: src/Microsoft.AspNetCore.Mvc.Core/Internal/FileResultExecutorBase.cs [hunk 6]
```diff
@@ -80,24 +70,35 @@ internal enum PreconditionState
...
+                SetAcceptRangeHeader(context);
+                // Assuming the request is not a range request, the Content-Length header is set to the length of the entire file. 
+                // If the request is a valid range request, this header is overwritten with the length of the range as part of the 
+                // range processing (see method SetContentLength).
+                response.ContentLength = fileLength.Value;
+                if (HttpMethods.IsHead(request.Method) || HttpMethods.IsGet(request.Method))
                 {
-                    serveBody = false;
-                    response.StatusCode = StatusCodes.Status412PreconditionFailed;
+                    if ((preconditionState == PreconditionState.Unspecified ||
```
Comment (Tratcher) [inline src/Microsoft.AspNetCore.Mvc.Core/Internal/FileResultExecutorBase.cs]: If you add the returns above I don't think you need to check preconditionState again here.

**Comment by jbagga** [comment]
Reverting last commit. Making the suggested change means that the Accept Range header will not be set even if file has a length because the method returns before that code.

File: src/Microsoft.AspNetCore.Mvc.Core/Internal/FileResultExecutorBase.cs (hunk 5 above)
Comment (jbagga) [inline src/Microsoft.AspNetCore.Mvc.Core/Internal/FileResultExecutorBase.cs]: Returning here would mean that this is never reached https://github.com/aspnet/Mvc/pull/6348/files#diff-bb6a01f81a4cd717b9350677639fd1f5R87

**Pull request merged**